shared/cluster.go

```diff
@@ -409,3 +409,139 @@ func WriteDataPod(namespace string) (string, error) {
 
 	return RunCommandHost(cmd)
 }
+
+type ServiceType string
+
+const (
+	Server ServiceType = "server"
+	Agent  ServiceType = "agent"
+)
+
+func (s ServiceType) IsValidServiceType() bool {
+	valid := []ServiceType{Server, Agent}
+	for _, i := range valid {
+		if s == i {
+			return true
+		}
+	}
+
+	return false
+}
+
+type ServiceAction string
+
+const (
+	Stop    ServiceAction = "stop"
+	Start   ServiceAction = "start"
+	Restart ServiceAction = "restart"
+	Status  ServiceAction = "status"
+)
+
+func (s ServiceAction) IsValidServiceAction() bool {
+	valid := []ServiceAction{Stop, Start, Restart, Status}
+	for _, i := range valid {
+		if s == i {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (s ServiceAction) isStartAction() bool {
+	startAction := []ServiceAction{Start, Restart}
+	for _, i := range startAction {
+		if s == i {
+			return true
+		}
+	}
+
+	return false
+}
+
+type Product string
+
+const (
+	K3S  Product = "k3s"
+	RKE2 Product = "rke2"
+)
+
+func (p Product) IsValidProduct() bool {
+	valid := []Product{K3S, RKE2}
+	for _, i := range valid {
+		if p == i {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (p Product) getString() string {
+	if p == K3S {
+		return "k3s"
+	}
+
+	return "rke2"
+}
+
+// getServiceName Get service name. Used to work with stop/start k3s/rke2 services
+func (product Product) getServiceName(serviceType ServiceType) string {
+	if !serviceType.IsValidServiceType() {
+		ReturnLogError("serviceType is not valid; has to be set to ONLY: server | agent")
+	}
+	var serviceName string
+	if product == K3S && serviceType == Server {
+		serviceName = product.getString() // k3s
+	} else { // k3s-agent, rke2-server, rke2-agent
+		serviceName = fmt.Sprintf("%s-%s", product, serviceType)
+	}
+
+	return serviceName
+}
+
+func (product Product) GetServiceCmd(action ServiceAction, serviceType ServiceType) string {
+	if action.isStartAction() {
+		return fmt.Sprintf("timeout 2m sudo systemctl --no-block %s %s; sleep 60", action, product.getServiceName(serviceType))
+	}
+	return fmt.Sprintf("timeout 2m sudo systemctl --no-block %s %s", action, product.getServiceName(serviceType))
+}
+
+// ManageClusterService action:stop/start/restart/status product:rke2/k3s ips:ips array for serviceType:agent/server
+func (product Product) ManageClusterService(action ServiceAction, serviceType ServiceType, ips []string) error {
+	if !product.IsValidProduct() {
+		ReturnLogError("product needs to be one of: k3s | rke2")
```

Comment:
https://github.com/rancher/distros-test-framework/pull/42/files#r1397405156

no need for this , since in the beginning of your call stack u are loading the product from here , which is loading from  https://github.com/rancher/distros-test-framework/blob/dbc0224ca28dba1bd9567c33a0a89112ed2904bf/config/reader.go#L33C2-L54C1 

![Screenshot 2023-11-17 at 11 33 47](https://github.com/rancher/distros-test-framework/assets/124721967/a938c561-675b-4faf-9fe2-ea1629865bf3)

Reply:
removed all product validations.